scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/TaskState.java

```diff
@@ -86,12 +94,24 @@ public String toString() {
     /*
      * Creates an initial state for tasks in a given |generation|.
      *
-     * Such initial state starts at the beginning of the generation and spans for
-     * |windowSizeMs| milliseconds.
+     * Such initial state starts at the beginning of the generation and spans
+     * until now minus confidence window size.
      */
-    public static TaskState createInitialFor(GenerationId generation, long windowSizeMs) {
-        Timestamp generationStart = generation.getGenerationStart();
-        return new TaskState(generationStart, generationStart.plus(windowSizeMs, ChronoUnit.MILLIS), Optional.empty());
+    public static TaskState createInitialFor(GenerationId generation, Timestamp now,
+                                             long confidenceWindowSizeMs, long queryTimeWindowSizeMs) {
+        // Start reading at generation start:
+        Timestamp windowStart = generation.getGenerationStart();
+
+        // Create a large window up to (now - confidenceWindowSizeMs), except
+        // when the window gets too small - in that case create a window
+        // queryTimeWindowSizeMs large (the consumer might need to wait a bit
+        // for the window to be ready for reading).
+        Timestamp windowEnd = now.plus(-confidenceWindowSizeMs, ChronoUnit.MILLIS);
+        if (windowEnd.compareTo(windowStart) < 0) {
```

Comment:
Is this enough? Wouldn't we want to adjust windowEnd also when `windowEnd - windowStart < queryTimeWindowSizeMs`?

Reply:
Maybe, but this is only an initial window so it doesn't really matter.

Reply:
I could add this at the cost of additional complexity.